csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1788,21 +1825,198 @@ class StackBasedSharedMemAllocator : kir::IrVisitor {
 
 } // namespace
 
+// Use allocation info map to find aliases, i.e. allocations that are properly
+// sized and parallelized so that they can be re-used without any
+// synchronization.
+std::vector<Expr*> aliasMemoryAllocations(
+    const std::vector<Expr*>& exprs,
+    AllocationInfoMap& allocation_info_map) {
+  ReusableAllocationFinder::find(exprs, allocation_info_map);
+  return AllocationAliasModifier::modify(exprs, allocation_info_map);
+}
+
+class PromoteReuseSyncModifier : private kir::ExprMutator {
+ public:
+  PromoteReuseSyncModifier(
+      const std::vector<Expr*>& exprs,
+      const AllocationInfoMap& allocation_info_map)
+      : allocation_info_map_(allocation_info_map) {
+    // Find next allocation after last aliased read of all allocations whose
+    // reuse we need to promote, and record shortest sync intervals relative to
+    // subsequent allocations.
+    for (const auto& alloc_info : allocation_info_map.allAllocationInfos()) {
+      auto tv = alloc_info->alloc_expr->buffer()->as<TensorView>();
+      if (tv->getMemoryType() != MemoryType::Shared || !tv->getPromoteReuse()) {
+        continue;
+      }
+      auto last_read = alloc_info->getAliasedOuterLastRead();
+
+      std::optional<int> nearest_first_write = std::nullopt;
+
+      for (const auto& other : allocation_info_map.allAllocationInfos()) {
+        auto first_write = other->outer_live_interval->firstWrite();
+        if (first_write <= last_read) {
+          continue;
+        }
+        if (!nearest_first_write.has_value() ||
+            first_write < nearest_first_write.value()) {
+          nearest_first_write = first_write;
+        }
+      }
+
+      if (nearest_first_write.has_value()) {
+        sync_intervals_.emplace(last_read, nearest_first_write.value());
+      }
+    }
+
+    if (sync_intervals_.empty()) {
+      exprs_ = exprs;
+    } else {
+      if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
+        debug() << "Ensuring syncs within these intervals:" << std::endl;
+        for (auto [last_read, first_write] : sync_intervals_) {
+          debug() << "  " << last_read << " - " << first_write << std::endl;
+        }
+      }
+      traverseAndInsert(exprs);
+    }
+  }
+
+  const std::unordered_set<Expr*>& insertedSyncs() const {
+    return inserted_syncs_;
+  }
+
+  const std::vector<Expr*>& modifiedExprs() const {
+    return exprs_;
+  }
+
+ private:
+  using kir::ExprMutator::dispatch;
+
+  void dispatch(Expr* expr) final {
+    // Skip inserted syncs
+    if (inserted_syncs_.find(expr) != inserted_syncs_.end()) {
+      if (isDebugDumpEnabled(DebugDumpOption::BufferReuseInfo)) {
+        debug() << "Skipping new sync expression " << expr->toString();
+      }
+      kir::ExprMutator::dispatch(expr);
+    }
```

Comment:
IIUC, `traverseAndInsert` will only insert after the entire traverse is done, which means, inside here, we will never see an inserted sync? (If it did insert on the fly, then should we recompute `AllocationInfoMap` every time when we register an insertion?)

Reply:
You're right it should not insert until afterward; this check is not needed. I had gotten a segfault at one point and placed this guard there to diagnose, but I think that was a logic error that wound up calling `kir::ExprMutator::dispatch`. I'll verify that none of the tests are hitting it and remove if not.